File: LeetCode/509-fibonacci-number/509-fibonacci-number.py
from dataclasses import dataclass


class Solution:
    def fib(self, n: int) -> int:
        # f[0] = 0
        # f[1] = 1
        # f[2] = f[0] + f[1]
        # f[3] = f[1] + f[2]
        f = []
        f.append(0)
        f.append(1)

        for i in range(2, n+1):
            f.append(f[i-2]+f[i-1])

        return f[n]


@dataclass
class Quest():
    inp: int
    sol: int


def checkAns(q: Quest) -> bool:
    sol = Solution()
    out = sol.fib(q.inp)
    res = (q.sol == out)
    if not res:
        print(f'{q.inp} -> {q.sol} != {out}')
    return res


if __name__ == '__main__':
    q = Quest(inp=2, sol=1)
    checkAns(q)

    q = Quest(inp=3, sol=2)
    checkAns(q)

    q = Quest(inp=4, sol=3)
    checkAns(q)

File: LeetCode/509-fibonacci-number/509-fibonacci-number.c
#include <stdio.h>

int fib(int n)
{
    int f[31] = {0};
    f[0] = 0;
    f[1] = 1;
    for (int i = 2; i <= n; i++)
        f[i] = f[i - 2] + f[i - 1];
    return f[n];
}

struct quest
{
    int inp;
    int sol;
};
typedef struct quest Quest;

int checkAns(Quest q)
{
    int out = fib(q.inp);
    int res = out == q.sol;

    if (!res)
    {
        printf("%d -> %d != %d\n", q.inp, q.sol, out);
    }

    return res;
}

int main()
{
    Quest q;

    q = (Quest){2, 1};
    checkAns(q);

    q = (Quest){3, 2};
    checkAns(q);

    q = (Quest){4, 3};
    checkAns(q);

    q = (Quest){10, 55};
    checkAns(q);

    return 0;
}